Remove "context saved" expectation from CoreDataStack mocks

crazytonyli opened the pull request:
This PR is part of paaHJt-3ky-p2.

## Changes
Remove `testExpectation` from `TestContextManager` and `ContextManagerMock`. This expectation is used to assert `ContextManager.saveContext` method are called. This logic is split into two areas: created in unit test method, but fulfilled outside of test method. This PR moves all this logic into unit test method, for better readability, it also simplifies `CoreDataStack` mock implementations.

## Test Instructions
Make sure all unit tests pass.

## Regression Notes
N/A. All changes in this PR are in unit test, they don't have any impact on the app.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

crazytonyli committed:
Remove "context saved" expectation from CoreDataStack mocks

They are moved to test methods.

crazytonyli commented on WordPress/WordPressTest/ContextManagerMock.m:
```diff
@@ -79,40 +76,13 @@ - (NSManagedObjectContext *)mainContext
+    // FIXME: Remove this method to use superclass one instead
+    // This log magically resolves a deadlock in
+    // `ZDashboardCardTests.testShouldNotShowQuickStartIfDefaultSectionIsSiteMenu`
+    NSLog(@"Context save completed");
```
I'm still trying to figure out cause of this deadlock 👀

mokagio replied:
For what is worth, I can see the deadlock without this line on my end, too. The test you mention just sits there with a spinner next to it in the test navigator.

mokagio commented on WordPress/WordPressTest/BlogJetpackTest.m:
```diff
@@ -75,19 +75,17 @@ - (void)testJetpackUsername {
     AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:[ContextManager sharedInstance].mainContext];
     WPAccount *wpComAccount = [accountService createOrUpdateAccountWithUsername:@"user" authToken:@"token"];
-    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+    [self waitForExpectations:@[saveExpectation] timeout:2.0];
```
I noticed this PR replaces all the `waitForExpectations(timeout:, handler:)` with `wait(for:, timeout:)`.

What's your rationale for this change?

crazytonyli replied:
I believe this expectation was the only expectation created within this test method. This change is just making that explicit - test method is waiting for main context to be saved.

mokagio replied:
Fair enough. One could argue that using `waitForExpectations(timeout:, handler:)` would make for smaller diffs, because if we add new expectations that method will automatically wait for them.

But, I'm with you on the value of explicitly stating which expectation we're waiting for. 👍

mokagio commented on WordPress/WordPressTest/BlogJetpackTest.m:
```diff
@@ -75,19 +75,17 @@ - (void)testJetpackUsername {
 - (void)testJetpackSetupDoesntReplaceDotcomAccount {
-    XCTestExpectation *saveExpectation = [self expectationWithDescription:@"Context save expectation"];
-    self.testContextManager.testExpectation = saveExpectation;
+    XCTestExpectation *saveExpectation = [self expectationForNotification:NSManagedObjectContextDidSaveNotification object:self.testContextManager.mainContext handler:nil];
 
     AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:[ContextManager sharedInstance].mainContext];
```
🤔 it's surprising to see the expectation using `self.textContextManager.mainContext` and this using `[ContextManager sharedInstance].mainContext`.

They look like different contexts. Or am I missing something?

For reference, here's where each is instantiated (I think)

For `ContextManager`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/112b352938ecbea191addd94be7f61d8cd4d5bf4/WordPress/Classes/Utility/ContextManager.m#L72-L83

For `TestContextManager`, which internally uses `ContextManager`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/112b352938ecbea191addd94be7f61d8cd4d5bf4/WordPress/WordPressTest/ContextManagerMock.m#L70-L80

crazytonyli replied:
Good catch! The code does look like they are referencing different context objects. But at runtime, they are actually the same one, the one in `ContextManagerMock`, since TestContextManager [overrides the shared instance](https://github.com/wordpress-mobile/WordPress-iOS/blob/98dc3b777d8e2cc5f82b66b5e2a2b92153e74e3c/WordPress/WordPressTest/TestContextManager.m#L15-L16)(the overriding actually happens [within `ContextManagerMock`](https://github.com/wordpress-mobile/WordPress-iOS/blob/98dc3b777d8e2cc5f82b66b5e2a2b92153e74e3c/WordPress/WordPressTest/ContextManagerMock.m#L23)). Which is why `self.testContextManager.mainContext` is same as `[[ContextManager sharedInstance] mainContext]`.

I will update the code to use the same reference for clarity.

BTW, the above overriding behaviour will be updated in my establishing usage pattern PR.

mokagio replied:
Gotcha! I knew I must have been missing something, else I would have expected the tests to fail 😄 

Thank you for making it explicit in https://github.com/wordpress-mobile/WordPress-iOS/pull/18485/commits/c6873f2c15b02d4593c31565dac87e2924e85058. 👍

mokagio reviewed (commented):
This is shaping up well, thank you @crazytonili!

The deadlock behavior, and it's resolution with a simple log, is quite baffling... 😳 

I'd say there's something deeper down in the CoreData stack and/or how the Dashboard part of the app uses it. I'm saying this because the tests were explicitly prefixed with `Z` to let them run at the end.

See:

- https://github.com/wordpress-mobile/WordPress-iOS/commit/3a3e1aee4fa1206fbcc9d8f30336c980489ea3a0
- https://github.com/wordpress-mobile/WordPress-iOS/blob/112b352938ecbea191addd94be7f61d8cd4d5bf4/WordPress/WordPressTest/Dashboard/BlogDashboardServiceTests.swift

Also, `BlogDashboardServiceTests` explicitly mentions CoreData as a possible cause of issues:

https://github.com/wordpress-mobile/WordPress-iOS/blob/112b352938ecbea191addd94be7f61d8cd4d5bf4/WordPress/WordPressTest/Dashboard/BlogDashboardServiceTests.swift#L6-L13

You shared [a possible fix](https://github.com/crazytonyli/WordPress-iOS/commit/aca7dabdfe2b4c034138df18ea4c944a29e9480b#diff-2fe4dfa10c8e9b74a4ecdf11a6ec23b485a01a3132435c259f820bc0f19c06e0) and I _love_ your initiative. But, I'd say let's focus on the test suite for the context of this trial because we want to keep it time constrained.

IMHO, having a `FIXME` comment plus `NSLog` is no worse than renaming tests to start with `Z` so they run last as a workaround.

crazytonyli committed:
Use `testContextManager` instead of `ContextManager.sharedInstance`

They are the same object in runtime, since `TestContextManager`
overrides `ContextManager.sharedInstance` with a mock instance.

mokagio reviewed (approved):
I think this is now good to merge. Thank you @crazytonyli, and congrats on your first WordPress iOS contribution 😄 

I updated https://github.com/wordpress-mobile/WordPress-iOS/pull/18488 so it will run the tests in CI for your latest changes. If that's green, we should be good to merge the PR 🎉

mokagio commented:
@crazytonyli 

> If that's green, we should be good to merge the PR 🎉 

Actually, as an external contributor, you wouldn't be able to merge the PR. I will enable auto-merge, so if/when the tests pass, this will land on `trunk`.

Also notice I added labels and milestones. That's a bit of our process that's not _required_ during the trial but would still be appropriate. 

_Rule of thumb to choose the correct milestone value: if at the start of a sprint, use the current milestone (i.e. the one that will close at the start of the next sprint); if towards the end of a sprint (say only 2-3 days left) use the next milestone._

_In our case, today is Tuesday, and it's the first week of the sprint. 19.8 is the milestone in code freeze (see the ❄️  in the name),  19.9 is the current milestone (the one that we'll close at the start of next sprint), and 20.0 is the next milestone (the WordPress apps adhere to the [WordPress versioning](https://codex.wordpress.org/WordPress_Versions) which is not strictly SemVer: once we reach `x.9` we roll to `(x+1).0`_

crazytonyli commented:
🥳 Thanks for merging this PR!

I did notice the labels and milestone in other PRs. But I don't think I have the permission to add labels nor milestone.

mokagio commented:
> But I don't think I have the permission to add labels nor milestone.

Oh, I wasn't expecting that... No worries, then. Someone will add them
